### PR TITLE
Refactor QueryStringQuery for 6.0

### LINKED ISSUE #25574: Unify `query_string` free text parsing and `multi_match` parsing
We discussed the removal of the long standing `split_on_whitespace` issue for the `query_string`
 with @jpounz and @colings86 and decided that 6.0 is the right time to introduce a major change in `query_string` behavior:
**Disable the ability to split_on_whitespace entirely.**

This proposal goes further than #25470 by unifying the parsing experience between the `match` family query and the `query_string` parser but also to force users to not use a mode that relies on whitespace splitting. `match_query` and `muti_match_query` do not have the concept of `split_on_whitespace` so instead we want to align `query_string` to use the same mechanism than a `multi_match` query but applied only on the "text part" separated by operators.
For instance the query string `"foo bar AND foobar"` has two "free text" parts: 

- `foo bar` 
- and `foobar`.  

Changing the behavior requires also to adapt the other options to the new rule.
The following is the list of required changes for `query_string` to split text on operators only without loosing any features from the ancient mode:

- [ ]  Ignore `split_on_whitespace` and always split on operators instead.

- [ ]  Logs a deprecation warning when `split_on_whitespace` is set

- [ ] Introduce a new option named `type` that applies the https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#multi-match-types on each free text part of the `query_string`.

- [ ] Makes the new option `type` mutually exclusive with `use_dismax` and `auto_generate_phrase_queries`

- [ ]  Make`use_dismax:true` query equivalent to`type:best_fields` or `type:most_fields` depending on the value of the `tiebreaker` option.

 - [ ] Make `auto_generate_queries:true` equivalent to `match_phrase` type. 

 - [ ] Document the new behavior and adds a section regarding the behavior of `min_should_match`
 coupled with the new `type`option.

### PR DESCRIPTION
**Pull request by jimczi**
This change refactors the `query_string` query to analyze the query text around logical operators of the query string the same way than a `match_query`/`multi_match_query`.
It also adds a `type` parameter that can be used to change the way multi fields query are built the same way than a `multi_match` query does.

Now that these queries share the same behavior regarding text analysis, some parameters are obsolete and have been deprecated:

 * `split_on_whitespace`: This setting is now ignored with a deprecation notice
      if it is used explicitely. With this PR The `query_string` always splits on logical operator.
      It simplifies the understanding of the other parameters that can have different meanings 
      depending on the value of  `split_on_whitespace`.

 * `auto_generate_phrase_queries`: This setting is now ignored with a deprecation notice
      if it is used explicitely. This setting only makes sense when the parser splits on whitespace.

* `use_dismax`: This setting is now ignored with a deprecation notice
     if it is used explicitely.  The `tie_breaker` parameter is sufficient to handle `best_fields`/`most_fields`.

Fixes #25574

### CODE REVIEW

**Comment by jimczi** [comment]
Document is still a WIP but the code is ready. It's a big PR though, I tried to separate the commits a bit but one is still much bigger, sorry for the reviewer ^^ @colings86

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java [hunk 1]
```diff
@@ -609,16 +616,19 @@ public boolean escape() {
     }
 
     /**
-     * Whether query text should be split on whitespace prior to analysis.
-     * Default is <code>{@value #DEFAULT_SPLIT_ON_WHITESPACE}</code>.
+     * This query parser splits on operator only.
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Maybe explicitly state that this setting is ignored like `autoGeneratePhraseQueries`?

**Review by colings86 — COMMENTED** [review]
LGTM for the code. I know you said the docs are still WIP but just a quick note that I think we should add deprecated notes to the options in the query string query topic as well as mentioning it in the breaking changes. Maybe also include int eh not that the setting is ignored.

**Comment by jimczi** [comment]
I pushed a commit to handle the `type` option in the `query_string` query and adds the documentation for the new behavior.
I think it is ready, @colings86 can you take another look ?

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java [hunk 2]
```diff
@@ -110,6 +111,7 @@
     private static final ParseField SPLIT_ON_WHITESPACE = new ParseField("split_on_whitespace")
             .withAllDeprecated("The parser splits on logical operator");
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: Should we explicitly say here that this option has no effect?

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java [hunk 3]
```diff
@@ -989,8 +1012,11 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
         }
 
         queryParser.setDefaultOperator(defaultOperator.toQueryParserOperator());
+        queryParser.setType(type);
         if (tieBreaker != null) {
             queryParser.setGroupTieBreaker(tieBreaker);
+        } else {
+            queryParser.setGroupTieBreaker(type.tieBreaker());
```
Comment (colings86) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: minor: Should we set this in `setType()` in the `queryParser` since the value is tied to the type anyway?

**Review by colings86 — APPROVED** [review]
LGTM: I left two minor comments but don't feel strongly on them so feel free to push with or without them

**Comment by jimczi** [comment]
jenkins test this please

File: core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java (hunk 3 above)
Comment (jimczi) [inline core/src/main/java/org/elasticsearch/index/query/QueryStringQueryBuilder.java]: The `tie_breaker` overrides the value embedded in `type` so it is not needed. We always set the `tie_breaker` in the multi match query builder when it's needed.

**Comment by Bargs** [comment]
@jimczi I was playing around with these changes in master today and they seem a bit confusing.

If I'm understanding things correctly, it seems like we're being inconsistent with our requirement of logical operators. 

The query string `foo bar` will search all fields for the text `foo bar`, passing that exact text to each field's analyzer.

But `message:foo bar` will search the message field for `foo` and all fields for `bar`. Once a field is introduced, we go back to splitting on whitespace. Instead, I would expect this to search the message field for `foo bar`. Similarly, I would expect `message:foo otherField:bar` to search the message field for `foo otherField:bar`. To create an actual bool query, the user should have to explicitly use an `AND`/`OR` e.g. `message:foo AND otherField:bar`.

In other words, when I read "with this PR the query_string always splits on logical operator" I understand that to mean "the query_string never splits on whitespace". I think it would be ideal if that were the case. I've seen many Kibana users get confused by query_string's whitespace splitting.

Let me know what you think or if I'm misunderstanding something.

**Comment by jimczi** [comment]
Thanks for testing @Bargs !

> In other words, when I read "with this PR the query_string always splits on logical operator" I understand that to mean "the query_string never splits on whitespace"

Sorry the description is misleading, I meant `the query_string always splits around operators`. The `logical` part is an abuse of mine that I corrected in:
https://github.com/elastic/elasticsearch/commit/4a9995145c359678a50db019e5d0a90e7d064256

That means that operators defined in the parser syntax are used as delimiter for the analyzer. The named field query (`field:`)  is an operator as such as the prefix query operator (`*`) and many others:
https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-syntax
I hope that this clarifies how things work internally, IMO it's the right compromise as it makes the output predicable.

> Similarly, I would expect message:foo otherField:bar to search the message field for foo otherField:bar

I don't think we should do that. Splitting on named field makes it simpler to read and if you want to have the multi words analysis you can use parenthesis:
`message:(new york city)`

**Comment by thomassuckow** [comment]
Does this mean that in 6.0

```json
"query_string": {
  "query": "new yo* city",
  "default_field": "myfield",
  "analyze_wildcard": true
}
```

Would do-what-you-expect™ in that it would search for terms in order with a wildcard in the middle?

**Comment by mainameiz** [comment]
> split_on_whitespace

Does it work as in 5.x versions for fields of `text` type?

**Comment by mainameiz** [comment]
https://www.elastic.co/guide/en/elasticsearch/reference/6.x/query-dsl-query-string-query.html

> Whitespaces are not considered operators, this means that `new york city` will be passed "as is" to the analyzer configured for the field. If the field is a keyword field the analyzer will create a single term `new york city` and the query builder will use this term in the query. If you want to query each term separately you need to add explicit operators around the terms (e.g. new AND york AND city).

What means this:

> If the field is a keyword field

It works differently for `text` field?
